### PR TITLE
Migrate to go1.22

### PR DESCRIPTION
**Pull request by dusk125**
Migrate main to golang 1.22.

One task towards https://github.com/etcd-io/etcd/issues/17348

Ran some benchmarks between this branch and main locally as follows:
```
# Server
GOMAXPROCS=10 ./bin/etcd --quota-backend-bytes=17179869184 --listen-client-urls http://0.0.0.0:23790 --advertise-client-urls http://127.0.0.1:23790

# Benchmark
GOMAXPROCS=6 go run ./tools/benchmark put --conns=64 --clients=64 --key-size=256 --val-size=4096 --key-space-size=50000 --total=2000000 --endpoints "http://127.0.0.1:23790"

# make clean && make build between each run
```

|   | Run 1  | Run 2  | Run 3  |
|---|---|---|---|
| main<br>etcd Version: 3.6.0-alpha.0<br>Go Version: go1.21.6<br>Go OS/Arch: linux/amd64  | Summary:<br>Total:        631.2755 secs.<br>Slowest:      0.1076 secs.<br>Fastest:      0.0028 secs.<br>Average:      0.0202 secs.<br>Stddev:       0.0092 secs.<br>Requests/sec: 3168.1889  | Summary:<br>Total:        679.5380 secs.<br>Slowest:      0.1487 secs.<br>Fastest:      0.0027 secs.<br>Average:      0.0217 secs.<br>Stddev:       0.0097 secs.<br>Requests/sec: 2943.1761  | Summary:<br>Total:        669.9843 secs.<br>Slowest:      0.1206 secs.<br>Fastest:      0.0026 secs.<br>Average:      0.0214 secs.<br>Stddev:       0.0097 secs.<br>Requests/sec: 2985.1447  |
| this branch<br>etcd Version: 3.6.0-alpha.0<br>Go Version: go1.22.0<br>Go OS/Arch: linux/amd64  | Summary:<br>Total:        651.7652 secs.<br>Slowest:      0.1282 secs.<br>Fastest:      0.0033 secs.<br>Average:      0.0209 secs.<br>Stddev:       0.0099 secs.<br>Requests/sec: 3068.5899  | Summary:<br>Total:        647.3450 secs.<br>Slowest:      0.1279 secs.<br>Fastest:      0.0030 secs.<br>Average:      0.0207 secs.<br>Stddev:       0.0097 secs.<br>Requests/sec: 3089.5426  | Summary:<br>Total:        699.5499 secs.<br>Slowest:      0.1271 secs.<br>Fastest:      0.0033 secs.<br>Average:      0.0224 secs.<br>Stddev:       0.0097 secs.<br>Requests/sec: 2858.9813  |

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -2,23 +2,21 @@
 // README at: https://github.com/devcontainers/templates/tree/main/src/go
 {
 	"name": "Go",
-	
 	// Or use a Dockerfile or Docker Compose file. More info: https://containers.dev/guide/dockerfile
-	"image": "mcr.microsoft.com/devcontainers/go:1.21-bookworm",
-
+	"image": "mcr.microsoft.com/devcontainers/go:1.20-bookworm",
```
Comment (ahrtr) [inline .devcontainer/devcontainer.json:6]: Why change this?

Also did you manually update this file or was it automatically updated by any tool?

Comment (dusk125) [reply .devcontainer/devcontainer.json:6]: That was a mistake, I meant to update it to 1.22 as well. I've fixed this in the most recent patch

**Review by serathius — APPROVED** [review]

**Comment by ahrtr** [comment]
I see there are some issue against go 1.22, let's on-hold bumping 1.22 for a while.

- https://github.com/golang/go/issues?q=is%3Aissue+is%3Aopen+22+in%3Atitle
- also failed in bbolt on Windows

File: .go-version [hunk 2]
```diff
@@ -1 +1 @@
-1.21.6
+1.22.0
```
Comment (ahrtr) [inline .go-version:1]: Please bump to 1.22.1, thanks

**Comment by ahrtr** [comment]
We need to bump google.golang.org/protobuf to 1.33.0 to resolve the CVE
 https://github.com/protocolbuffers/protobuf-go/releases/tag/v1.33.0

**Comment by ahrtr** [comment]
We have to merge this PR firstly, and let https://github.com/etcd-io/etcd/pull/17543 to take care of bumping google.golang.org/protobuf to 1.33.0

**Review by ahrtr — APPROVED** [review]

**Pull request merged**